### ISSUE TITLE
Allow custom root ca to be injected in all sidecars

### ISSUE THREAD

**Issue by florianmutter** [istio/istio#52437]
**Describe the feature request**
When using ServiceEntry to access external services a custom root ca might be required to validate the TLS connection. The only way we could find to get this working is to add a volume mount to the proxy with the pem certificate and specify the path to it in the DestinationRule.
In clusters with a lot of deployments this produces a lot of duplicate code in all deployments that use istio which we we would like to avoid. We also need to find all deployment yamls of all services that will get istio injected and add either the annotations (`sidecar.istio.io/userVolume` and `sidecar.istio.io/userVolumeMount`) to mount to volume or add the istio-proxy container with the volume.

It would be nice if istio could inject custom root ca certificates to all sidecar proxies creates based on the mesh config or some other centralized way.

**Describe alternatives you've considered**
* Setting `meshConfig.caCertificates` -> does not work
* Setting `meshConfig.defaultConfig.caCertificatesPem` -> does not work
* Create a custom injector template with the volume -> not great to maintain. Needs a config map in every namespace that has istio sidecars
* Create a custom mutating webhook to add the volume -> order of webhook execution is not guaranteed so it has to handle running before or after istio sidecar injector which makes it hard to maintain. Needs a config map in every namespace that has istio sidecars
* Create a custom base image with the certificate -> We would need to force everyone to use this. Our developers use a variety of ways to build images. We would like to add the certificate transparently to developers.
* Set `insecureSkipVerify` to `true` -> Insecure

**Affected product area (please put an X in all that apply)**

[ ] Ambient
[ ] Docs
[ ] Dual Stack
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[x] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Affected features (please put an X in all that apply)**

[ ] Multi Cluster
[ ] Virtual Machine
[ ] Multi Control Plane

**Additional context**

With Istio 1.21 istio validates certificates of external services by default. https://istio.io/latest/news/releases/1.21.x/announcing-1.21/upgrade-notes/#default-value-of-the-feature-flag-verify_cert_at_client-is-set-to-true

**Comment by howardjohn** [comment]
> Create a custom base image with the certificate -> We would need to force everyone to use this.

This would be in the image of the istio proxy not the application which may make it more viable?

https://kubernetes.io/docs/concepts/storage/projected-volumes/#clustertrustbundle fixes this problem pretty much, with the massive caveat that its only alpha in k8s 1.29, making it impractical at the moment

**Comment by florianmutter** [comment]
> This would be in the image of the istio proxy not the application which may make it more viable?

You are right but it seems this would also be done by changing all deployments. This also requires a lot of changes and we would have to establish processes to keep this up to date.

> [kubernetes.io/docs/concepts/storage/projected-volumes#clustertrustbundle](https://kubernetes.io/docs/concepts/storage/projected-volumes/#clustertrustbundle) fixes this problem pretty much, with the massive caveat that its only alpha in k8s 1.29, making it impractical at the moment

That looks like what we want. But this enhancement does not seem to be activly developed anymore so we might need to wait long for it.

I would actually like istio to do this it possible. It seems like there is already a lot of certificate handling and distribution in place there is just non that puts a new CA somewhere where envoy can find it. I'm a little bit confused though why the two options to set certs in mesh config do not work. The documentation says they are only for workload-to-workload but I could not understand from the code why it does not work for workload to external service as well.

For `caCertificatesPem` it looks to me like it would do what I want but is only read if dynamically pushed to the pod not if added to the default config (as far as I understand). The only place where it is read that I could find is here: https://github.com/istio/istio/blob/a0f07792c2ad2ececa27b5e1a212248b4305208c/pkg/istio-agent/xds_proxy.go#L174

**Comment by howardjohn** [comment]
> You are right but it seems this would also be done by changing all deployments. This also requires a lot of changes and we would have to establish processes to keep this up to date.

**Comment by nischay30** [comment]
We are also facing the same issue

**Comment by Sana7H** [comment]
We also faced this issue while upgrading to Istio 1.21 from 1.19. Can we get an update on this ticket please?

**Comment by zirain** [comment]
cc @leosarra @kfaseela

**Comment by Sana7H** [comment]
@howardjohn Let's say we are creating a custom proxy image with the certificate pre-loaded for external service TLS validation, what should be the path we should be using to load the certificate? /etc/certs?
From the documentation ` **Updated** default value of the feature flag `VERIFY_CERT_AT_CLIENT` to true.
  This means server certificates will be automatically verified using the OS CA certificates when not using a DestinationRule caCertificates field.`, what path does OS CA certificates here refer to?

**Comment by Sana7H** [comment]
I went through the code to understand where Istio reads the CA cert from https://github.com/istio/istio/blob/master/pkg/security/security.go#L452 , I'm going to try building a proxy image adding custom root ca into the expected path and test it out

**Comment by pmozdzynski** [comment]
Istiod  automatically detects a secret named cacerts in the istio-system namespace and mount it. Best feature will be to have same approach on the sidecars( (isto-proxy). However that is not possible yet and workaround are required. The best would be to get istio-proxy to inherit what istiod has.

**Comment by Max-LAB** [comment]
Lets reopen this again, I also have an interest here. Good that I found this. In short this is the problem we have.

On `Cluster-A`, we have Keycloak deployed, and it is accessed by applications in `Cluster-B` via the domain *.keycloak.xyz.com.

To enable communication from `Cluster-B` to Keycloaks `Cluster-A`, I manually add the `keycloak.xyz.com.crt` (signed by internal PKI) into the `/etc/ssl/certs` folder inside the application’s Docker image. After doing so, I can successfully use `curl` to reach the Keycloak service from within the application pod.

Problem is that this setup is not scalable. There will be multiple applications across different clusters, all needing access to Keycloak via `*.keycloak.xyz.com`.

I need a way to automatically distribute thekeycloak.xyz.com.crt certificate to all pods in the clusters that are communicating via Istio, without having to manually import the certificate into each application’s container. The ideal solution would allow Istio (via the Envoy proxy) to manage and distribute this certificate to all relevant pods across clusters.

Here is also one Q&A where I tried to find solution https://github.com/istio/istio/discussions/56196

**Comment by Max-LAB** [comment]
> The only way we could find to get this working is to add a volume mount to the proxy with the pem certificate and specify the path to it in the DestinationRule.

@florianmutter can you please write in here how your `ServiceEntry` and `DestinationRule` looks like for external service? I set my for external keycloak and on Keycloak cluster I am receivin always this:

```sh
[2025-05-14T12:34:35.415Z] "- - HTTP/1.1" 400 DPE http1.codec_error - "-" 0 11 0 - "-" "-" "-" "-" "-" - - 100.64.0.39:443 100.64.2.1:17821 keycloak.xyz.com - 
```
